---
title: "Impute z-statistics from PGC3 SCZ"
subtitle: 'European samples'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
    self_contained: true
params:
  dataset: NULL
  variable_type: NULL
  ctst_key: NULL
  AnnoLevel: NULL
  SampleLevel: NULL
---


<!---
cd /sc/arion/projects/roussp01a/gabriel/ref_panels/imputez_analysis

R

system("git pull")

rmarkdown::render("c4_ref.Rmd")

--->

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```


# C4 reference
```{sh download, eval=FALSE}
# ASSEMBLY=GRCh38 
ASSEMBLY=GRCh37

wget https://personal.broadinstitute.org/giulio/panels/MHC_haplotypes_CEU_HapMap3_ref_panel.${ASSEMBLY}.vcf.gz
bcftools norm -m -  MHC_haplotypes_CEU_HapMap3_ref_panel.${ASSEMBLY}.vcf.gz | bgzip > MHC_haplotypes_CEU_HapMap3_ref_panel.${ASSEMBLY}_norm.vcf.gz
tabix -p vcf MHC_haplotypes_CEU_HapMap3_ref_panel.${ASSEMBLY}_norm.vcf.gz

FILE=MHC_haplotypes_CEU_HapMap3_ref_panel.${ASSEMBLY}_norm.vcf.gz
echo -e "CHROM\tPOS\tID\tA1\tA2" > $(basename $FILE .vcf.gz).map
bcftools view $FILE | grep -v "#" | cut -f1-5 >> $(basename $FILE .vcf.gz).map
```


# Load data
```{r load}
library(imputez)
library(tidyverse)
library(GenomicDataStream)

# SCZ GWAS
# GRCh37
file = "/sc/arion/projects/roussp01a/gabriel/ref_panels/data/PGC3_SCZ_wave3.european.autosome.public.v3.vcf.tsv.gz"
df_z_obs = read_tsv(file, comment="##") %>%
    arrange(CHROM, POS) %>%
    mutate(z = BETA / SE)  %>%
    rename(GWAS_A1 = A1, GWAS_A2 = A2)

# C4 reference map
file = "/sc/arion/projects/roussp01a/gabriel/ref_panels/c4/MHC_haplotypes_CEU_HapMap3_ref_panel.GRCh37_norm.map"
df_map = read_delim(file)
colnames(df_map)[4:5] = c("REF_A1", "REF_A2")

# df_map[nchar(df_map$REF_A2) != 1,]

df = df_z_obs %>% 
  select(-CHROM, -POS) %>% 
	inner_join(df_map, by="ID")

# C4 reference VCF
file = "/sc/arion/projects/roussp01a/gabriel/ref_panels/c4/MHC_haplotypes_CEU_HapMap3_ref_panel.GRCh37_norm.vcf.gz"
gds = GenomicDataStream(file, field="GT", region="6", init=TRUE)

# impute
window = 10000000
flankWidth = 2500000
res = run_imputez(df, gds, window, flankWidth)	
```

# Process results
```{r results}
res_c4 = res %>%
			filter(ID == "C4") %>%
			mutate(ID = paste(ID, A1, A2, sep="__")) 

baseline = "<BS>"
i = which(res_c4$A2 == baseline)
z_base = res_c4$z.stat[i]

res_c4 = res_c4 %>%
	mutate(z = z_base - z.stat) %>%
	mutate(pvalue = 2*pnorm(abs(z), lower.tail=FALSE))

# get beta and se
# https://www.biostars.org/p/319584/
# Beta = z / sqrt(2p(1− p)(n + z^2)) and
# SE = 1 / sqrt(2p(1− p)(n + z^2))

# Get variance of each allele
gds = GenomicDataStream(file, field="GT", region="6", init=TRUE)
gds = reinitializeStream(gds)
dat = getNextChunk( gds )
dat$info$ID <- with(dat$info, paste(ID, A1, A2, sep="__"))
colnames(dat$X) <- dat$info$ID

var_x = apply(dat$X, 2, var) 

# var(dat$X[,1])
# p = sum(dat$X[,1]) / (2*nrow(dat$X))
# 2*p*(1-p)

i = match(names(var_x), res_c4$ID)
res_c4$var_x = NA
res_c4$var_x[i[!is.na(i)]] = var_x[!is.na(i)]


n = 58749.13
res_c4$beta = with(res_c4, z.stat / sqrt(var_x * (n + z.stat^2)))
res_c4$se = with(res_c4, 1 / sqrt(var_x * (n + z.stat^2)))


baseline = "<BS>"
i = which(res_c4$A2 == baseline)
beta_base = res_c4$beta[i]
res_c4$beta = beta_base - res_c4$beta


ord = c('BS' = 1,
'AL-BS-1' = 2,
'AL-BS-2' = 2,
'AL-BS-3' = 2,
'AL-BS-4' = 2,
'AL-BS-5' = 2,
'AL-BS-BS' = 2,
'AL-BS-other' = 2,
'AL'  = 2,
'AL-BL-1' = 3,
'AL-BL-2' = 3,
'AL-BL-3' = 3,
'AL-BL-BL' = 3,
'AL-BL-BS' = 3,
'AL-BL-other' = 3,
'AL-AL-1' = 4,
'AL-AL-2' = 4,
'AL-AL-3' = 4,
'AL-AL-AL' = 4,
'AL-AL-BL' = 4,
'AL-AL-BS' = 4,
'AL-AS-BL' = 4,
'AL-AS-BS-BS' = 4,
'BL' = 5,
'BL-BS' = 5)

cols = c("blue", "green2", "orange", "red2", "yellow2")
cols = cols[ord]
names(cols) = names(ord)

res_c4 %>%
	mutate(A2 = gsub("<", "", A2)) %>%
	mutate(A2 = gsub(">", "", A2)) %>%
	mutate(A2 = factor(A2, rev(names(ord)))) %>%
	ggplot(aes(beta, A2, color=A2)) +
		geom_errorbar(aes(xmin = beta - 1.96*se, xmax = beta + 1.96*se), width=0) + 
		geom_point() +
		theme_classic() +
		theme(aspect.ratio=1, legend.position="none") +
		geom_vline(xintercept=0, color="grey70", linetype="dashed") +
		scale_color_manual(values = cols) +
		xlab(bquote(beta))

# location of C4 alleles
loc = df_map %>% 
	filter(ID == "C4") %>% 
	pull(POS) %>%
	unique

ymax = -log10(min(res_c4$pvalue))*1.03

df %>%
	ggplot(aes(POS/1000000, -log10(PVAL))) +
		geom_point() +
		theme_classic() +
		theme(aspect.ratio=1/2) +
		ylab(bquote(-log[10]~P)) +
		xlab("chr6") +
		geom_point(data=res_c4, aes(loc/1000000, -log10(pvalue)), color="red") +
		scale_y_continuous(limits=c(0, ymax), expand=c(0,0)) +
		scale_x_continuous(expand=c(0,0))
```



```{r compare}
methods = c("decorrelate", "Schafer-Strimmer", "Ledoit-Wolf", "OAS", "Touloumis")

df_time = list()

res = lapply(methods, function(method){
	message(method)
	tm = system.time({
		res <- run_imputez(df, gds, window, flankWidth, method=method)	
		})
	df_time[[method]] <<- tm
	res 
})

colMethods = c("decorrelate" = "red",
              "GIW-EB (k=50)" = "#9f1214",
              "lambda = 0" = "#0000cd", 
              "Ledoit-Wolf" = "#FF7F00",    
              "OAS"= "#cccc29", 
              "Touloumis" = "#A65628", 
              "Schafer-Strimmer" = "#f569b3", 
              "Pseudoinverse" = "green3",
              "Baseline" = "grey50",
              "Oracle" = "black")

df_time = do.call(rbind, df_time) %>%
						as.data.frame %>%
						rownames_to_column('Method')

ymax = max(df_time$elapsed)*1.10

df_time %>%
	ggplot(aes(Method, elapsed, fill=Method, label=round(elapsed, digits=1))) +
		geom_bar(stat="identity") +
		theme_classic() +
		theme(aspect.ratio=1, legend.position="none") +
		scale_fill_manual(values = colMethods) +
		scale_y_continuous(limits=c(0, ymax), expand=c(0,0)) +
		ylab("Wall time (seconds)") +
		geom_text(aes(Method, elapsed+1), hjust=0) +
		coord_flip() 
```












