---
title: "Impute z-statistics from PGC3 SCZ"
subtitle: 'European samples'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
output: 
  html_document:
    toc: true
    smart: true
    self_contained: true
params:
  dataset: NULL
  variable_type: NULL
  ctst_key: NULL
  AnnoLevel: NULL
  SampleLevel: NULL
---


<!---


cd /sc/arion/projects/roussp01a/gabriel/ref_panels/imputez_analysis


system("git pull")


rmarkdown::render("imputez_PGC3_SCZ.Rmd")


--->

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```

# Load packages
```{r pkg, cache=FALSE}
library(imputez)
library(tidyverse)
library(GenomicDataStream)
library(ggplot2)
library(scattermore)
library(parallel)
library(RhpcBLASctl)
omp_set_num_threads(4)
```

# Load data
```{r load}
# Load names of samples to be included from reference panel
# FAM file
file = "/sc/arion/projects/data-ark/Public_Unrestricted/1000G/phase3/PLINK/chr1.fam"
df_fam = read.table(file)
colnames(df_fam) = c("FID", "IID", "PID", "MID", "Sex", "Pop")

# Population assignments
file = "/sc/arion/projects/data-ark/Public_Unrestricted/1000G/phase3/release_2013502/integrated_call_samples.20130502.ALL.ped"
df_pop = read.table(file, header=TRUE, sep="\t")
df_pop = df_pop[df_pop$Relationship %in% c("father", "unrel","mother"),]
df_pop = df_pop[df_pop$Individual.ID %in% df_fam$IID,]

pops = c('GBR','IBS', 'TSI','CEU')#, "FIN")
sampleIDs = df_pop$Individual.ID[df_pop$Population %in% pops]
length(sampleIDs)

# read variant positions
files = dir("/sc/arion/projects/roussp01a/gabriel/ref_panels/1kg/", pattern=".*_norm_eur.map", full.names=TRUE)
df_map = read_delim(files, progress=FALSE)
colnames(df_map)[4:5] = c("REF_A1", "REF_A2")
 
# GRCh37
file = "/sc/arion/projects/roussp01a/gabriel/ref_panels/data/PGC3_SCZ_wave3.european.autosome.public.v3.vcf.tsv.gz"
df_z_obs = read_tsv(file, comment="##") %>%
    arrange(CHROM, POS) %>%
    mutate(z = BETA / SE)  %>%
    rename(GWAS_A1 = A1, GWAS_A2 = A2)

# join
df = df_z_obs %>%
  select(-CHROM, -POS) %>% 
  inner_join(df_map, by="ID")

# if alleles are flipped, take negative z-stat
idx = with(df, GWAS_A1 == REF_A2 & GWAS_A2 == REF_A1)
tmp = df$GWAS_A1[idx]
df$GWAS_A1[idx] = df$GWAS_A2[idx]
df$GWAS_A2[idx] = tmp
df$z[idx] = df$z[idx]

# Keep variants where GWAS and REF panel alleles match
df = df %>%
  filter(GWAS_A1 == REF_A1)%>%
  filter(GWAS_A2 == REF_A2)

# Remove 
#-------
# Variant IDs with multiple alleles
exclID = with(df, unique(ID[duplicated(ID)]))

# Exclude MHC: "6:24894177-33890574"
exclID_MHC = df %>%
  filter(CHROM == 6, POS >= 24894177 & POS <= 33890574 ) %>%
  select(ID)

df = df %>%
    filter(!ID %in% c(exclID, exclID_MHC)) %>%
    filter( IMPINFO > 0.95) 
```


```{r impute}
window = 1000000
flankWidth = 250000

methods = c("decorrelate", "GIW-EB (k=100)", "Schafer-Strimmer", "OAS", "Touloumis", "Pseudoinverse") #, "Ledoit-Wolf"

# for each method
res = mclapply( methods, function(method){
  tm <- system.time({
    # for each chrom
    res = lapply( 22, function(chrom){
      message(paste(method, chrom))

      # 1kg reference file
      file = paste0("/sc/arion/projects/roussp01a/gabriel/ref_panels/1kg/1kg_chr", chrom, "_norm_eur.bcf")
      gds = GenomicDataStream(file, field="GT", init=TRUE, samples=sampleIDs, MAF=0.01)

      # filter
      df_chrom = df[df$CHROM == chrom,]

      # exclude a set of variants
      idx = seq(1, nrow(df_chrom), by=10)

      # run imputation
      if( method == "GIW-EB (k=100)"){
        res = run_imputez(df_chrom[-idx,], gds, window, flankWidth, method="decorrelate", k=100)
      }else if( method == "Pseudoinverse"){        
        res = run_imputez(df_chrom[-idx,], gds, window, flankWidth, method="S", lambda=0, useginv=TRUE)
      }else{        
        res = run_imputez(df_chrom[-idx,], gds, window, flankWidth, method=method)
      }
      res$method = method

      # # merge with observed z-statistics
      res %>%
        inner_join( df_chrom[idx,], by="ID")
    }) 
    })
  outfile = paste0("./", method, ".RDS")
  saveRDS(tm, outfile)
  bind_rows(res)
}, mc.cores=6)
df_res = bind_rows(res)

files = dir("./", pattern="*.RDS")
df_time = lapply(files, readRDS)
names(df_time) = gsub(".RDS", "", basename(files))
```


```{r time, cache=FALSE}
colMethods = c("decorrelate" = "red",
              "GIW-EB (k=50)" = "#9f1214",
              "lambda = 0" = "#0000cd", 
              "Ledoit-Wolf" = "#FF7F00",    
              "OAS"= "#cccc29", 
              "Touloumis" = "#A65628", 
              "Schafer-Strimmer" = "#f569b3", 
              "Pseudoinverse" = "green3",
              "Baseline" = "grey50",
              "Oracle" = "black")

rmse = function(x) sqrt(mean(x^2))

# Time
do.call(rbind, df_time) %>%
  as.data.frame %>%
  rownames_to_column('Method') %>%
  ggplot(aes(Method, elapsed, fill=Method)) +
    geom_bar(stat="identity") +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_fill_manual(values = colMethods) +
    scale_y_continuous(limits=c(0, NA), expand=c(0,0)) +
    ylab("Wall time (seconds)") +
    coord_flip()

do.call(rbind, df_time) %>%
  as.data.frame %>%
  rownames_to_column('Method') %>%
  ggplot(aes(Method, user.self, fill=Method)) +
    geom_bar(stat="identity") +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_fill_manual(values = colMethods) +
    scale_y_continuous(limits=c(0, NA), expand=c(0,0)) +
    ylab("CPU time (seconds)") +
    coord_flip()

MAF = 0.05

# rMSE
df_res %>%
  filter(maf > MAF) %>%
  group_by(method) %>%
  summarize(rMSE = rmse(z.stat - z), 
      rMSE.mod = rmse(z.stat/se - z)) %>%
  ggplot(aes(method, rMSE, fill=method)) +
    geom_bar(stat="identity") +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_fill_manual(values = colMethods) +
    scale_y_continuous(limits=c(0, NA), expand=c(0,0)) +
    ylab("Root mean squared error")

# rMSE - stratified
df_res %>%
  filter(maf > MAF) %>%
  group_by(method, nVariants) %>%
  summarize(rMSE = rmse(z.stat - z), 
      rMSE.mod = rmse(z.stat/se - z)) %>%
  ggplot(aes(nVariants, rMSE, color=method)) +
    geom_point() +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_color_manual(values = colMethods) +
    scale_y_continuous(limits=c(0, NA), expand=c(0,0)) +
    ylab("Root mean squared error")



# maf vs r2.pred
df_res %>%
  filter(maf > MAF) %>%
  ggplot(aes(maf, r2.pred, color=method)) +
    geom_scattermore(pointsize=3) +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_color_manual(values = colMethods) +
    facet_wrap( ~ method)

# maf vs error
df_res %>%
  filter(maf > MAF) %>%
  ggplot(aes(maf, abs(z.stat - z), color=method)) +
    geom_scattermore(pointsize=3) + 
    # geom_smooth() +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    # facet_wrap( ~ method) +
    scale_color_manual(values = colMethods) 

df_res %>%
  filter(maf > MAF) %>%
  ggplot(aes(maf, abs(z.stat - z), color=method)) +
    geom_smooth() +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_color_manual(values = colMethods) 

# maf vs error
breaks = c(0.01, 0.05, .1, .2, .3, .4, .5)
df_res %>%
  filter(maf > MAF) %>%
  mutate(cat = cut(maf, breaks)) %>%
  ggplot(aes(cat, abs(z.stat - z), color=method)) +
    geom_boxplot() +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    # facet_wrap( ~ method) +
    scale_color_manual(values = colMethods) +
    coord_cartesian(ylim=c(0, 1))

# maf vs delta
a = df_res %>%
  filter(method == "decorrelate") %>%
  mutate(delta = abs(z.stat - z))

b =  df_res %>%
  filter(method == "Schafer-Strimmer") %>%
  mutate(delta = abs(z.stat - z))

dfc = inner_join(a,b, by=c("ID", "A1", "A2"))

dfc %>%  
  filter(maf.x > MAF) %>%
  ggplot(aes(maf.x, delta.x - delta.y)) +
    geom_scattermore(pointsize=3) + 
    geom_hline(yintercept=0, color="grey70",linetype="dashed") +
    geom_smooth() +
    theme_classic() +
    theme(aspect.ratio=1)

dfc %>%
  arrange(delta.x) %>%
  tail(1) %>%
  data.frame

# rs139745380
dfc %>%
  arrange(delta.x - delta.y) %>%
  tail(1) %>%
  data.frame

dfc %>%  
  arrange(-maf.x) %>%
  filter(maf.x > 0.01) %>%
  ggplot(aes(delta.x,  delta.y, color=maf.x)) +
    geom_scattermore(pointsize=3) + 
    scale_color_gradient(low="blue", high="red", limits=c(0,.5)) +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="right") 

# rMSE - stratified by r2.pred
df_res %>%
  filter(maf > MAF) %>%
  ggplot(aes(r2.pred, (z.stat - z)^2, color=method)) +
    geom_point() +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_color_manual(values = colMethods) +
    scale_y_continuous(limits=c(0, NA), expand=c(0,0)) +
    ylab("Squared error")

df_res %>%
  filter(maf > MAF) %>%
  ggplot(aes(r2.pred, abs(z.stat - z), color=method)) +
    geom_point() +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    scale_color_manual(values = colMethods) +
    scale_y_continuous(limits=c(0, NA), expand=c(0,0)) +
    ylab("Absolute error")

df_res %>%
  filter(maf > MAF) %>%
  group_by(method) %>%
  summarize(rho = cor(r2.pred, abs(z.stat - z), method="sp"))

# imputez vs observed z-statistics
lim = range(c(df_res$z, df_res$z.stat))
df_res %>%
  filter(maf > 0.05) %>%
  arrange(-r2.pred) %>%
  ggplot(aes(z, z.stat, color=r2.pred)) +
    geom_scattermore(pointsize=7) +
    theme_classic() +
    xlab("Observed z-statistic") +
    ylab("Imputed z-statistic") +
    geom_abline(slope=1, intercept=0) +
    geom_hline(yintercept=0, color="grey70", linetype="dashed") +
    geom_vline(xintercept=0, color="grey70", linetype="dashed") +
    scale_color_gradient(low="grey30", high="red", limits=c(0,1)) +
    coord_fixed(ratio = 1) +
    ylim(lim) + 
    xlim(lim) +
    facet_wrap( ~ method)

df_res %>%
  filter(maf > 0.05) %>%
  arrange(-r2.pred) %>%
  ggplot(aes(z.stat, se, color=r2.pred)) +
    geom_point() +
    theme_classic() +
    theme(aspect.ratio=1) +
    scale_color_gradient(low="grey30", high="red", limits=c(.7,1)) +
    facet_wrap( ~ method)

# histogram of lambda values
df_res %>%
    ggplot(aes(lambda)) +
      geom_histogram() +
      theme_classic() +
      theme(aspect.ratio=1) +
      xlab(bquote(lambda)) +
      facet_wrap( ~ method) +
      xlim(0, 1)

# histogram of r2.pred values
df_res %>%
    ggplot(aes(r2.pred)) +
      geom_histogram() +
      theme_classic() +
      theme(aspect.ratio=1) +
      xlab(bquote(r2.pred)) +
      facet_wrap( ~ method) +
      xlim(NA, 1)
```


### SessionInfo
```{r sessioInfo, cache=FALSE}
sessionInfo()
```



